crates/project_panel/src/project_panel.rs

```diff
@@ -1259,28 +1259,113 @@ impl ProjectPanel {
             } else {
                 None
             };
-
-            cx.spawn(|this, mut cx| async move {
+            let next_selection = self.find_next_selection_after_deletion(cx);
+            cx.spawn(|panel, mut cx| async move {
                 if let Some(answer) = answer {
                     if answer.await != Ok(0) {
                         return Result::<(), anyhow::Error>::Ok(());
                     }
                 }
                 for (entry_id, _) in file_paths {
-                    this.update(&mut cx, |this, cx| {
-                        this.project
-                            .update(cx, |project, cx| project.delete_entry(entry_id, trash, cx))
-                            .ok_or_else(|| anyhow!("no such entry"))
-                    })??
-                    .await?;
+                    if let Some(task) = panel
+                        .update(&mut cx, |panel, cx| {
+                            panel
+                                .project
+                                .update(cx, |project, cx| project.delete_entry(entry_id, trash, cx))
+                        })
+                        .ok()
+                        .flatten()
+                    {
+                        task.await.ok();
+                    }
                 }
+
+                panel.update(&mut cx, |panel, cx| {
+                    if let Some(next_selection) = next_selection {
+                        panel.selection = Some(next_selection);
+                        panel.autoscroll(cx);
+                    } else {
+                        panel.select_last(&SelectLast {}, cx);
+                    }
+                })?;
                 Result::<(), anyhow::Error>::Ok(())
             })
             .detach_and_log_err(cx);
             Some(())
         });
     }
 
+    fn find_next_selection_after_deletion(
```

Comment:
Overall, sounds somewhat heavy as we're iterating over all `entries` in the project panel.
That might be thousands of files easily.

Can we be smarter?
We have a `marked_entries()` method that will return all entries that will be deleted.

We could get all their entries in ~O(1) each and 
1. Do the sanitization I've been talking about before, removing all entries that are subfiles/subdirs (this subset can be determined way above this, in the removal method)
2. Then, we can start looking for entries that are adjacent to the last removed entry: presumably, we can find a parent dir, and then reverse-iterate this subset only, finding the new selection target.

Reply:
Addressed above.